python/lsst/ctrl/bps/bps_config.py

```diff
@@ -150,6 +141,14 @@ def search(self, key, opt=None):
             Key to look for in config
         opt: `dict`, optional
             Options to use while searching
+                curvals: `dict`, optional
+                    Means to pass in values for search order key (curr_<sectname>) or variable replacements
+                default: optional
+                    Value to return if not found
+                replaceVars: `bool`, default = True
+                    If search result is string, whether to replace variables inside it
+                required: `bool`, default = False
+                    If replacing variables, whether to raise exception if variable is undefined
```

Comment:
As they say, no good deed goes unpunished...

In my opinion, this entire docstring would benefit from being rewritten and expanded, preferably with some examples. With the `opt` option, the functionality of `search()` method goes way beyond mere checking whether a setting was defined in a configuration and what's its value. Yet figuring out the details is close to impossible without looking directly at the function definition itself.

Having said that, I'm fine with doing this work as part of some other, clean-up ticket. However, in this pull request, please address the following issues:

1. The type of the `default` variable is not specified (line146).
2. LSST Dev Guide suggests slightly different formatting when describing a dictionary type (lines 144-151). For details follow this [link](https://developer.lsst.io/python/numpydoc.html#dictionary-types).
3. The term *hierarchy rules* is rather "cryptic", *search order* may be better option.

Finally, while it is mostly a matter of personal taste, I'm not a huge fan of using abbreviations as "drop-in" replacements for actual words in sentences, proper or not. Words like *val* or *opt* works just fine as a variable name, but reading phrases like *current vals*, *given opt* in the description feels rather, I don't know, awkward.

Reply:
Leaving code examples in the docstring to some other ticket.  Otherwise making these changes.